File: src/main/java/me/zeroeightsix/kami/module/modules/chat/AutoQMain.java
package me.zeroeightsix.kami.module.modules.chat;

import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import net.minecraft.network.play.client.CPacketChatMessage;

import java.text.SimpleDateFormat;
import java.util.Date;

import static me.zeroeightsix.kami.util.MessageSendHelper.sendChatMessage;
import static me.zeroeightsix.kami.util.MessageSendHelper.sendWarningMessage;

/**
 * @author dominikaaaa
 * Updated by d1gress/Qther on 5/12/2019
 * Updated by dominikaaaa on 26/03/20
 */
@Module.Info(
        name = "AutoQMain",
        description = "Automatically does '/queue main' on servers",
        category = Module.Category.CHAT,
        showOnArray = Module.ShowOnArray.OFF
)
public class AutoQMain extends Module {

    private Setting<Boolean> showWarns = register(Settings.b("Show Warnings", true));
    private Setting<Boolean> connectionWarning = register(Settings.b("Connection Warning", true));
    private Setting<Boolean> dimensionWarning = register(Settings.b("Dimension Warning", true));
    private Setting<Double> delay = register(Settings.doubleBuilder("Wait time").withMinimum(0.2).withValue(7.1).withMaximum(10.0).build());

    private double delayTime;
    private double oldDelay = 0;

    @Override
    public void onUpdate() {
        if (mc.player == null) return;
        if (oldDelay == 0) oldDelay = delay.getValue();
        else if (oldDelay != delay.getValue()) {
            delayTime = delay.getValue();
            oldDelay = delay.getValue();
        }

        if (delayTime <= 0) {
            delayTime = (int) (delay.getValue() * 2400);
        } else if (delayTime > 0) {
            delayTime--;
            return;
        }

        if (mc.getCurrentServerData() == null && connectionWarning.getValue()) {
            sendMessage("&l&6Error: &r&6You are in singleplayer");
            return;
        }
        if (!mc.getCurrentServerData().serverIP.equalsIgnoreCase("2b2t.org") && connectionWarning.getValue()) {
            sendMessage("&l&6Warning: &r&6You are not connected to 2b2t.org");
            return;
        }
        if (mc.player.dimension != 1 && dimensionWarning.getValue()) {
            sendMessage("&l&6Warning: &r&6You are not in the end. Not running &b/queue main&7.");
            return;
        }
        sendQueueMain();
    }

    private void sendQueueMain() {
        SimpleDateFormat formatter = new SimpleDateFormat("HH:mm:ss");
        Date date = new Date(System.currentTimeMillis());
        System.out.println(formatter.format(date));
        sendChatMessage("&7Run &b/queue main&7 at " + (formatter.format(date)));
        mc.playerController.connection.sendPacket(new CPacketChatMessage("/queue main"));
    }

    private void sendMessage(String message) { if (showWarns.getValue()) sendWarningMessage(getChatName() + message); }

    public void onDisable() { delayTime = 0; }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/chat/AutoQMain.kt
package me.zeroeightsix.kami.module.modules.chat

import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.ModuleConfig.setting
import me.zeroeightsix.kami.util.TickTimer
import me.zeroeightsix.kami.util.TimeUnit
import me.zeroeightsix.kami.util.text.MessageSendHelper
import me.zeroeightsix.kami.util.text.MessageSendHelper.sendServerMessage
import me.zeroeightsix.kami.util.threads.safeListener
import net.minecraftforge.fml.common.gameevent.TickEvent
import java.text.SimpleDateFormat
import java.util.*

@Module.Info(
    name = "AutoQMain",
    description = "Automatically does '/queue 2b2t-lobby'",
    category = Module.Category.CHAT,
    showOnArray = false
)
object AutoQMain : Module() {
    private val showWarns = setting("ShowWarnings", true)
    private val dimensionWarning = setting("DimensionWarning", true)
    private val delay = setting("Delay", 30, 5..120, 5)

    private val timer = TickTimer(TimeUnit.SECONDS)

    init {
        safeListener<TickEvent.ClientTickEvent> {
            if (!timer.tick(delay.value.toLong())) return@safeListener

            if (mc.currentServerData == null) {
                sendMessage("&l&6Error: &r&6You are in singleplayer")
                return@safeListener
            }

            if (!mc.currentServerData!!.serverIP.equals("2b2t.org", ignoreCase = true)) {
                return@safeListener
            }

            if (player.dimension != 1 && dimensionWarning.value) {
                sendMessage("&l&6Warning: &r&6You are not in the end. Not running &b/queue main&7.")
                return@safeListener
            }

            sendQueueMain()
        }
    }

    private fun sendQueueMain() {
        val formatter = SimpleDateFormat("HH:mm:ss")
        val date = Date(System.currentTimeMillis())

        MessageSendHelper.sendChatMessage("&7Run &b/queue 2b2t-lobby&7 at " + formatter.format(date))
        sendServerMessage("/queue 2b2t-lobby")
    }

    private fun sendMessage(message: String) {
        if (showWarns.value) MessageSendHelper.sendWarningMessage("$chatName $message")
    }
}
